### ISSUE TITLE
NPE on requestOneTimePayment

### ISSUE THREAD

**Issue by nzakharchenko** [braintree/braintree_android#69]
After releasing application with included Braintree I start getting such error http://crashes.to/s/bae120cb021. 

```compile "com.braintreepayments.api:braintree:2.1.0"```

```minSdkVersion 14```
```targetSdkVersion 23```
```compileSdkVersion 23```
```buildToolsVersion "23.0.2"```

```
@Override
    public void onConfigurationFetched(Configuration configuration) {
        PayPalRequest request = new PayPalRequest(new BigDecimal(EntityManager.get().getCartTotalPrice()).setScale(2, BigDecimal.ROUND_HALF_UP).toString());
        request.currencyCode(Reservation.get().currency().getLabel());

        PayPal.requestOneTimePayment(mBraintreeFragment, request);
    }
```

**Comment by quinnjn** [comment]
Hmm interesting, that stack trace is weird. 

> com.braintreepayments.api.PayPal.requestOneTimePayment(PayPal.java:1212)

[PayPal](https://github.com/braintree/braintree_android/blob/master/BraintreeApi/src/main/java/com/braintreepayments/api/PayPal.java) isn't that big.

**Comment by sdcoffey** [comment]
@quinnjn Yeeah I was just looking at this, could be referring to an anonymous inner class in the full version of `requestOneTimePayment`

**Comment by quinnjn** [comment]
@nzakharchenko can you describe your integration in more detail?

> com.braintreepayments.api.PayPal.requestOneTimePayment(PayPal.java:1212)

This stack trace is really odd to us. 

We don't have that many LOC in PayPal.java, and nothing is sticking out to us in [requestOneTimePayment(BraintreeFragment, PayPalRequest)](https://github.com/braintree/braintree_android/blob/master/BraintreeApi/src/main/java/com/braintreepayments/api/PayPal.java#L158) that might cause the NPE.

**Comment by nzakharchenko** [comment]
So, after getting **clientToken** from our API, I'm creating **BraintreeFragment**
```
    @Override
    protected void afterPaymentStart(final PaymentStartResult paymentStartResult) {
        final boolean hasValidResult = !paymentStartResult.isError()
                && paymentStartResult.result
                && !TextUtils.isEmpty(paymentStartResult.client_token_id);

        if (!hasValidResult) {
            handleStartPaymentError();
        } else {
            mClientToken = paymentStartResult.client_token_id;

            try {
                mBraintreeFragment = BraintreeFragment.newInstance(this, mClientToken);
            } catch (InvalidArgumentException e) {
                // client token or client key was invalid
                handleStartPaymentError();
            }
        }
    }
```

Also I have implemented listener for **onConfigurationFetched**
```
    @Override
    public void onConfigurationFetched(Configuration configuration) {
        PayPalRequest request = new PayPalRequest(new BigDecimal(EntityManager.get().getCartTotalPrice()).setScale(2, BigDecimal.ROUND_HALF_UP).toString());
        request.currencyCode(Reservation.get().currency().getLabel());

        PayPal.requestOneTimePayment(mBraintreeFragment, request);
    }
```

And for **onPaymentMethodNonceCreated** where I'm actually saving **none** and sending it to our API
```
    @Override
    public void onPaymentMethodNonceCreated(final PaymentMethodNonce paymentMethodNonce) {
        if (paymentMethodNonce == null || TextUtils.isEmpty(paymentMethodNonce.getNonce())) {
            cancelPayment();
        } else {
            mNonce = paymentMethodNonce.getNonce();
            commitPayment();
        }
    }
```

@quinnjn Is it more clear now?

**Comment by quinnjn** [comment]
@nzakharchenko your `afterPaymentStart`, and `onPaymentMethodNonceCreated` sections look fine to me.

I was actually more interested in what you see for `PayPal.java:1212` since we can't see it. Are you able to give us more context on what is at this line number?

**Comment by lkorth** [comment]
@nzakharchenko: If you'd like to simplify things, there is no need to place your call to `PayPal#requestOneTimePayment` inside of `onConfigurationFetched` unless you want to check the configuration to see if PayPal is enabled. `PayPal#requestOneTimePayment` will correctly handle waiting for configuration if it is not yet available.

**Comment by nzakharchenko** [comment]
@lkorth does it actually means that I could call ```PayPal#requestOneTimePayment``` immediately after ```BraintreeFragment.newInstance(this, mClientToken);``` ? I think I've tried that and it was not working, but it was long time ago, maybe you've changed something.

**Comment by lkorth** [comment]
Yes, you can call it immediately after. If it doesn't work, let us know the error you are running into and we can take a look.

**Comment by nzakharchenko** [comment]
Ok! Thank you for pointing! I'll check and let you know

**Comment by nzakharchenko** [comment]
@lkorth I've tried to do that, and it works:
```
            try {
                mBraintreeFragment = BraintreeFragment.newInstance(this, mClientToken);

                PayPalRequest request = new PayPalRequest(new BigDecimal(EntityManager.get().getCartTotalPrice()).setScale(2, BigDecimal.ROUND_HALF_UP).toString());
                request.currencyCode(Reservation.get().currency().getLabel());

                PayPal.requestOneTimePayment(mBraintreeFragment, request);
            } catch (InvalidArgumentException e) {
                // client token or client key was invalid
                handleStartPaymentError();
            }
```

I'll release next version with this fix.

And about my crash in general - looks like it is some Fabric fuckup with line numbers, and to see correct line number you should just remove firs digit - so we will have ```212```.

Also I have a couple of other crashes, should I create separate issues for all of them?

**Comment by lkorth** [comment]
On `212` it's just `fragment.waitForConfiguration(new ConfigurationListener() {` so the only NPE I'd see would be if `null` was being passed to `PayPal#requestOneTimePayment` instead of the `BraintreeFragment`.

Yes, if you are seeing other crashes, please create an issue for each on and we'll take a look.

**Comment by nzakharchenko** [comment]
@lkorth yes, you are right, I'm also thinking that it could be ```null``` instead of ```BraintreeFragment```, but to be honest I could not imagine in which case it could happened. But let's see after next release with this changes that you suggested. I'll keep you updated.

**Comment by lkorth** [comment]
I'm also suspicious of the line numbers, one of the other crashes you created an issue for has `PayPal.java:5000`. Removing the most significant digit doesn't really work for that one :)

**Comment by nzakharchenko** [comment]
@lkorth yes, true.. for that one with line number ```5000``` it doesn't work, but for others - it works. I'm gonna contact with Fabric and ask about that problem

**Comment by lkorth** [comment]
@nzakharchenko we're currently working on several large changes related to where this crash occurred. These changes should go out in the next minor release. I'll be sure to keep these issues updated.

**Comment by lkorth** [comment]
@nzakharchenko: we recently released version 2.2.0 with several changes to the areas you are seeing crashes, please let us know if you are still seeing any crashes in the new version.
